Add PlatformDispatcher.engineId

knopp opened the pull request:
Fixes https://github.com/flutter/flutter/issues/163430.

This PR adds `engineId` field to `PlatformDispatcher`. When provided by the engine, this can be used to retrieve the engine instance from native code.

Dart code:
```dart
final identifier = PlatformDispatcher.instance.engineId!;
```

macOS, iOS: 
```objc
FlutterEngine *engine = [FlutterEngine engineForIdentifier: identifier];
```

Android:
```java
FlutterEngine engine = FlutterEngine.engineForId(identifier);
```

Linux
```cpp
FlEngine *engine = fl_engine_for_id(identifier);
```

Windows
```cpp
FlutterDesktopEngineRef engine = FlutterDesktopEngineForId(identifier);
```

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

knopp committed:
Add PlatformDispatcher.engineHandle to allow retrieving engine instance from native code

knopp committed:
Android implementation

javadoc

licenses

knopp committed:
iOS

knopp committed:
Linux

knopp committed:
Windows

matanlurey commented on dev/integration_tests/android_engine_test/test_driver/engine_handle_test.dart:
This is ... OK.

A better solution would be to use `package:integration_test`, and instead just write a (functional) unit test:

```dart
void main() {
  test('verify that...', () {
    expect(PlatformDispatcher.instance.engineHandle, equals(1), reason: 'Explain what "1" means.');
  });
}
```

However, this is probably not hooked up right now.

If you end up going with this solution, can you file an issue, and leave a TODO For `matanlurey`, and I'll address it?

knopp replied:
I looked into that, but it seems like that'd require some setup and also changes to test runner, I don't think I want to do this as part of this PR. Will happily leave the TODO though :)

matanlurey commented on dev/integration_tests/android_engine_test/test_driver/engine_handle_test.dart:
```diff
@@ -0,0 +1,30 @@
+  test('verify that engineHandle is set and works', () async {
+    final Map<String, Object?> response =
+        json.decode(await flutterDriver.requestData('')) as Map<String, Object?>;
+    expect(response['engineHandle'], 1);
```
Can you add a `reason: '...'` explaining what `1` means?

```dart
reason: 'A value of `1` means an engineHandle is registered.',
```

knopp replied:
Done.

matanlurey reviewed (commented):
For `dev/integration_tests/android_engine_test`, mostly LG, please address comment.

knopp committed:
Add TODO for matanlurey and explain the 1.

robert-ancell commented on engine/src/flutter/shell/platform/linux/fl_engine.cc:
```diff
@@ -34,6 +34,9 @@ static constexpr size_t kPlatformTaskRunnerIdentifier = 1;
 static constexpr int32_t kPointerPanZoomDeviceId = 1;
 
+static int64_t next_handle = 1;
+static GHashTable* engine_map;
```
Having these global values is a red flag - instead the handle should be set with `fl_engine_set_handle` and retrieved with `fl_engine_get_handle`. If we need to look up an engine it would probably live in the place that owns the engines, which is likely to be `FlApplication` (though we only support one engine at this point).

knopp replied:
The engine must be obtainable from handle without any other context. Requiring FlApplication instance in order to lookup the engine seems to negate the purpose of this PR?

robert-ancell replied:
Reading the linked issue, shouldn't the handle just be the memory address of the engine, and then this can used by the FFI directly? Then we don't need any IDs or maps.

knopp replied:
I want to avoid casting it to pointer directly and also detect invalid / outdated handles, so that we can ensure that `fl_engine_for_handle` will return `null` if the engine no longer exists.

knopp replied:
Alternatively, if we relax the requirements, `fl_engine_for_handle` implementation can be 
```cpp
g_return_val_if_fail(FL_IS_ENGINE(engine_handle), nullptr);
return FL_ENGINE(engine_handle)
```
I don't feel too strongly about this, if the global map and counter is too much of an issue.

knopp replied:
On other platforms we don't have similar checks though, so we'll probably still be using maps. Not sure if inconsistent behavior between platforms is a good trade-off.

robert-ancell replied:
The handle should remain valid until `FlutterEngineShutdown` is called. If the platform and UI thread is the same then we should always have safe interactions.

loic-sharma replied:
Personally, I also like Robert's suggestion of using the engine's memory address.

> ... so that we can ensure that `fl_engine_for_handle` will return `null` if the engine no longer exists.

While this would be a nice check, it doesn't completely solve the problem of a plugin holding to an engine pointer after it is destroyed.

Windows already has APIs like `ViewControllerFromHandle` that will gladly return garbage if you pass a handle that no longer exists.

> Not sure if inconsistent behavior between platforms is a good trade-off.

We could claim the behavior is consistent by stating `fl_engine_for_handle` has undefined behavior if the handle no longer exists? 🙃

knopp replied:
`ViewControllerFromHandle` simply casts `FlutterDesktopViewControllerRef` (C API) to `flutter:FlutterWindowsViewController` (c++ class). I don't think that's the same. In think case the "handle" is clearly a pointer. I think better comparison would be that once we have the ViewForId method - that will never return garbage for invalid Id. 

I think it all depends on the threading mdoel. Passing raw engine pointer as handle can be dangerous if the plugin needs to do the hop from UI to platform thread, because while the callback waits to be executed on platform thread the engine might be destroyed and the plugin has have no way of detecting that handle is invalid. 

This would be much less of an issue if we get the threading change in.

knopp replied:
@robert-ancell, this is the implementation for `fl_engine_for_id` now:
```cpp
FlEngine* fl_engine_for_id(int64_t id) {
  void* engine = reinterpret_cast<void*>(id);
  g_return_val_if_fail(FL_IS_ENGINE(engine), nullptr);
  return FL_ENGINE(engine);
}
```

robert-ancell reviewed (changes requested):
The term handle seems inappropriate. In the case of views we have an ID, and this seems to be the correct term to use here.

knopp commented:
I personally like the term handle better in this context, but I can see how it is not consistent with view identifier. I'll look into it.

loic-sharma commented on engine/src/flutter/shell/platform/windows/flutter_windows_engine.cc:
```diff
@@ -156,6 +160,9 @@ FlutterWindowsEngine::FlutterWindowsEngine(
   }
 
+  engine_handle_ = next_engine_handle_++;
+  handle_to_engine_[engine_handle_] = this;
```
My understanding is Windows lets you use different threads for different HWNDs. Should we have some sort of lock here?

knopp replied:
You can use HWND on the thread that it was created on (the thread must it's own message loop). All our HWNDs are created on main thread, and all embedder API assume main thread (same as other platforms). There's no need for locking. But we could maybe add thread affinity checks in future, at least for debug builds.

loic-sharma replied:
> All our HWNDs are created on main thread, and all embedder API assume main thread (same as other platforms).

I don't think that's true. The developer is free to edit their app entrypoint to launch their window and Flutter Windows from some background thread. And the embedder API doesn't require that the platform thread is the main thread either, the embedder API just defers to the embedder's platform task runner. The Flutter Windows platform task runner should work on background threads too.

knopp replied:
I think I see what you mean. Technically there is no "main thread" on Windows. Main thread simply refers to the thread on which the engine is created (and to which the task runner HWND belongs). In theory it is possible to create two engines, each one of different thread and then pump message loop on both threads. That should work.

In any case the map is gone from the PR, so there is no need for lock. But you are right, we shouldn't assume that platform thread is same thread for all engine instances on windows.

loic-sharma commented:
cc @stuartmorgan as you have experience with FFI plugins

knopp committed:
Rename EngineHandle to EngineId to be consistent with ViewId.

knopp committed:
No global engineId tracking

knopp commented:
I renamed `engineHandle` to `engineId` to not create inconsistency with `viewId`.

I also added a commit that removes global maps, so the `engineId` is a casted pointer. I'm not fully convinced about the tradeoff, but the global tracking of Ids seems to raise eyebrows and with platform thread merging hopefully landing in foreseeable future this shouldn't be much of an issue.

stuartmorgan-g reviewed (dismissed):
Per ongoing discussion in the linked issue, I don't think this should be exposed as a public embedding API until questions about what usage outside the SDK should look like, and what else we would need to implement to make that safe, have been resolved.

stuartmorgan commented on engine/src/flutter/shell/platform/darwin/ios/framework/Headers/FlutterEngine.h:
```diff
@@ -163,6 +163,17 @@ FLUTTER_DARWIN_EXPORT
+ *
+ * This function must be called on the main thread.
+ */
++ (nullable FlutterEngine*)engineForId:(int64_t)identifier;
```
Nit for either a private version, or for when this is added publicly: `engineForIdentifier:`

knopp replied:
Done.

knopp committed:
Rename engineForId: to engineForIdentifier:

knopp committed:
Make accessing instance from Id private for the time being

knopp commented:
While I don't think this open as big a can of worms as @stuartmorgan does, I understand the concern about rushing to introduce new public API for this. 

I added a commit to make all methods that provide engine for Id private, that way we can still experiment with dogfooding ffi within the embedders while decisions get made about how much of embedder API stability we want to guarantee.

This also contains enough functionality to fix `package:jni`, which could match the context with `engineId`, without providing access to the actual `FlutterEngine` from the Id.

stuartmorgan commented on engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine_Internal.h:
```diff
@@ -87,6 +87,22 @@ NS_ASSUME_NONNULL_BEGIN
+/**
+ * Returns the engine handle. Used in FlutterEngineTest.
+ */
+- (int64_t)engineId;
```
`engineIdentifier`

stuartmorgan-g reviewed (commented):
I didn't do a full review, but my concern about the public API is addressed, thanks!

knopp committed:
Rename

knopp committed:
Add missing @override annotation

loic-sharma commented on engine/src/flutter/lib/ui/platform_dispatcher.dart:
```diff
@@ -305,6 +305,13 @@ class PlatformDispatcher {
 
+  /// Opaque engine identifier for the engine running current isolate. Can be used
+  /// in native code to retrieve the engine instance.
+  /// The identifier is valid while the isolate is running.
```
It's not immediately obvious when `engineId` is null. Is that if the platform doesn't support engine IDs? Or is this null when the Dart isolate isn't running? Or during startup? This might be worth a quick explanation in the public docs.

loic-sharma commented on engine/src/flutter/shell/platform/embedder/embedder.cc:
```diff
@@ -2416,6 +2416,10 @@ FlutterEngineResult FlutterEngineInitialize(size_t version,
   }
 
+  if (args->engine_id != 0) {
+    run_configuration.SetEngineId(args->engine_id);
```
Should we call `run_configuration.SetEngineId` _before_ we check `run_configuration.IsValid`? This would ensure we generate the correct error if we ever decide to validate the run configuration's engine ID

knopp replied:
I'll move it.

loic-sharma commented on engine/src/flutter/shell/platform/embedder/embedder.cc:
```diff
@@ -2416,6 +2416,10 @@ FlutterEngineResult FlutterEngineInitialize(size_t version,
         "Could not infer the Flutter project to run from given arguments.");
   }
 
+  if (args->engine_id != 0) {
```
Should this be wrapped in a `SAFE_ACCESS`?

knopp replied:
It absolutely should. Thanks for catching this!

loic-sharma replied:
Would it be useful to have a test that launches the engine using a very old embedder API snapshot to verify we don't segfault?

knopp replied:
I think it'd be useful as regression test in follow-up PR if we actually merged this without the check. But I can still add it if you think it adds value.

loic-sharma commented on engine/src/flutter/shell/platform/windows/flutter_windows_internal.h:
```diff
@@ -65,6 +65,15 @@ FLUTTER_EXPORT void FlutterDesktopEngineRegisterPlatformViewType(
+//
+// This method must be called from the platform thread.
+FLUTTER_EXPORT FlutterDesktopEngineRef FlutterDesktopEngineForId(
+    int64_t engine_id);
```
Is it intentional that this is on the internal header file?

This is perfectly reasonable if you are concerned this API isn't production-ready yet... but I think it'd be OK to go straight to the public header file: `engine/src/flutter/shell/platform/windows/public/flutter_windows.h`.

Feel free to keep as-is, I'm just double-checking this was intentional :)

knopp replied:
@stuartmorgan is concerned that having this in public API opens a  can of worms, mostly if people start using bulk of the engine API from FFI, in which case we don't have good way to migrate the API without relying on runtime errors. You can read the whole conversion in attached issue. So for the time being the API stays private. It is still very useful to have this as we can start experiment with dogfooding FFI inside embedders.

loic-sharma commented on engine/src/flutter/shell/platform/embedder/embedder.h:
```diff
@@ -2631,6 +2631,11 @@ typedef struct {
+  /// Opaque identifier provided by the engine. Accessible in Dart code through
+  /// `PlatformDispatcher.instance.engineId`. Can be used in native code to
+  /// retrieve the engine instance that is running the Dart code.
+  int64_t engine_id;
```
@chinmaygarde Do you think we should create a `FlutterEngineId` typedef? Or is that over-kill?

For example, here's the typedef we have for `FlutterViewId`:

https://github.com/flutter/flutter/blob/50a1b953755b752f2a7823e381b41dbb54386cbd/engine/src/flutter/shell/platform/embedder/embedder.h#L277-L281

loic-sharma reviewed (approved):
`dart:ui`, engine, embedder API, and Windows changes look good to me!

Please also get approvals for Android, iOS, macOS, Linux, and web too

knopp committed:
Use SAFE_ACCESS when getting engineId. Move before check.

mdebbar reviewed (approved):
Web changes look good!

knopp committed:
Reformat because of trailing comma.

cbracken commented on engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:
```diff
@@ -241,6 +246,11 @@ - (instancetype)initWithName:(NSString*)labelPrefix
 
++ (FlutterEngine*)engineForIdentifier:(int64_t)identifier {
+  NSAssert([[NSThread currentThread] isMainThread], @"Must be called on the main thread.");
+  return (__bridge FlutterEngine*)reinterpret_cast<void*>(identifier);
```
This is sufficiently scary-looking as to draw the required attention :)

cbracken commented on engine/src/flutter/shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:
```diff
@@ -641,6 +641,7 @@ - (BOOL)runWithEntrypoint:(NSString*)entrypoint {
     }
     std::cout << message << std::endl;
   };
+  flutterArguments.engine_id = (int64_t)self;  // Conversion to intptr_t doesn't need bridged
```
Similar comment.

cbracken reviewed (approved):
![LGTM stamp from a Japanese personal seal](https://github.com/user-attachments/assets/b2eba158-654b-4702-9a4b-de7f4597434b)

lgtm modulo comments.

knopp commented on engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:
```diff
@@ -160,6 +160,11 @@ @implementation FlutterEngine {
 
+- (int64_t)engineIdentifier {
+  // Conversion to intptr_t doesn't need bridged cast.
+  return (int64_t)self;
```
I had that initially, but ARC reference [3.3.1](https://releases.llvm.org/6.0.0/tools/clang/docs/AutomaticReferenceCounting.html#id17) explicitly allows cast to `intptr_t`.  I can put it back, the lack symmetry bothers me too.

knopp committed:
Use bridged cast to raise eyebrows

knopp committed:
Add shell test for Null engineId

knopp commented on engine/src/flutter/shell/common/shell_unittests.cc:
```diff
@@ -5003,7 +5003,37 @@ TEST_F(ShellTest, SendViewFocusEvent) {
+  RunEngine(shell.get(), std::move(configuration));
+
+  latch.Wait();
+  ASSERT_EQ(reported_handle, 99);
```
Done.

robert-ancell reviewed (approved):
Linux embedder code LGTM

knopp committed:
safe access
